sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -384,22 +384,49 @@ object BooleanSimplification extends Rule[LogicalPlan] with PredicateHelper {
 /**
  * Simplifies binary comparisons with semantically-equal expressions:
  * 1) Replace '<=>' with 'true' literal.
- * 2) Replace '=', '<=', and '>=' with 'true' literal if both operands are non-nullable.
- * 3) Replace '<' and '>' with 'false' literal if both operands are non-nullable.
+ * 2) Replace '=', '<=', and '>=' with 'true' literal
+ *    if both operands are non-nullable or in constraints.
+ * 3) Replace '<' and '>' with 'false' literal if both operands are non-nullable or in constraints.
  */
-object SimplifyBinaryComparison extends Rule[LogicalPlan] with PredicateHelper {
+object SimplifyBinaryComparison
+  extends Rule[LogicalPlan] with PredicateHelper with ConstraintHelper {
+
+  private def canSimplifyComparison(
+      plan: LogicalPlan, left: Expression, right: Expression): Boolean = {
+    def canSimplify(left: Expression, right: Expression): Boolean = {
+      !left.nullable && !right.nullable && left.semanticEquals(right)
+    }
+
+    if (canSimplify(left, right)) {
+      return true
+    }
+
+    if (SQLConf.get.constraintPropagationEnabled) {
```

Comment:
Is this optimization constraint propagation only? When constraintPropagationEnabled is disabled, a Filter with predicate like `IsNotNull(e) && e > e` can still be simplified, no?

Reply:
Good catch! It also can optimize such`IsNotNull(e) && e > e` without `constraintPropagationEnabled`.